sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala

```diff
@@ -60,7 +60,13 @@ case class First(child: Expression, ignoreNulls: Boolean)
     this(child, FirstLast.validateIgnoreNullExpr(ignoreNullsExpr, "first"))
   }
 
-  override def nullable: Boolean = true
+  override def nullable: Boolean = {
+    if (ignoreNulls) {
+      false
```

Comment:
Hmm, even `ignoreNulls` is true to ignore null values, because `first`'s initial value is null, the evaluated value of `First` is still possibly null if,

1. No inputs, i.e. empty relation
2. All input values are null

I think this maybe why `nullable` is set to true always.

Although we can use `child.nullable` to check if input value could be null (2), but we cannot know if inputs are empty or not.

Reply:
All the input values are null is possible, e.g. `SELECT first(col) FILTER (col is null)`.

Reply:
@beliefer , I agree, as I added above, probably we just don't consider ignoreNulls and set first.nullable = child.nullable?

Reply:
> for the empty relation case, does it matter? because in the end we will have an empty table but with certain column as non-null defined in schema, I personally think there should be a difference between null value and the row doesn't exist at all

If the column is non-nullable in the empty relation, `First` will return null value but its `nullable` is false (because you set it to `child.nullable`). This produces inconsistent schema and data.

Reply:
I see what you mean now

yes, I reproduced it and it will even throw NPE in shuffle write stage

is there any disadvantage for us to change https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L549-L551 to `optimizedPlan.schema` ?

Reply:
I see what you mean now

yes, I reproduced it and it will even throw NPE in shuffle write stage

is there any disadvantage for us to change https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L549-L551 to `optimizedPlan.schema` ?